spring-integration-cloudevents/src/main/java/org/springframework/integration/cloudevents/transformer/ToCloudEventTransformer.java

```diff
@@ -0,0 +1,288 @@
+		ApplicationContext applicationContext = getApplicationContext();
+		if (this.sourceExpression == null) {  // in the case the user sets the value prior to onInit.
+			this.sourceExpression = new FunctionExpression<>((Function<Message<?>, URI>) message -> {
+				String appName = applicationContext.getEnvironment().getProperty("spring.application.name");
+				appName = appName == null ? "unknown" : appName;
+				return URI.create("/spring/" + appName + "." + getBeanName());
+			});
+		}
+		if (this.dataSchemaExpression == null) { // in the case the user sets the value prior to onInit.
+			this.dataSchemaExpression = new FunctionExpression<>((Function<Message<?>, @Nullable URI>)
+					message -> null);
+		}
+	}
+
+	/**
+	 * Transform the input message into a CloudEvent message.
+	 *
+	 * @param message the input Spring Integration message to transform
+	 * @return CloudEvent message in the specified format
+	 * @throws RuntimeException if serialization fails
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+
+		String id = this.idExpression.getValue(this.evaluationContext, message, String.class);
+		if (!StringUtils.hasText(id)) {
+			throw new MessageTransformationException(message, "No id was found with the specified expression");
+		}
+
+		URI source = this.sourceExpression.getValue(this.evaluationContext, message, URI.class);
+		if (source == null) {
+			throw new MessageTransformationException(message, "No source was found with the specified expression");
+		}
+
+		String type = this.typeExpression.getValue(this.evaluationContext, message, String.class);
+		if (type == null) {
+			throw new MessageTransformationException(message, "No type was found with the specified expression");
+		}
+
+		String contentType = message.getHeaders().get(MessageHeaders.CONTENT_TYPE, String.class);
+		if (contentType == null) {
+			throw new MessageTransformationException(message, "Missing 'Content-Type' header");
+		}
+
+		EventFormat eventFormat = this.eventFormatProvider.resolveFormat(contentType);
+		if (eventFormat == null) {
+			throw new MessageTransformationException("No EventFormat found for '" + contentType + "'");
+		}
+
+		ToCloudEventTransformerExtensions extensions =
+				new ToCloudEventTransformerExtensions(this.evaluationContext, (Message<byte[]>) message,
+						this.cloudEventExtensionExpressions);
+
+		CloudEvent cloudEvent = CloudEventBuilder.v1()
+				.withId(id)
+				.withSource(source)
+				.withType(type)
+				.withTime(OffsetDateTime.now())
+				.withDataContentType(contentType)
+				.withDataSchema(this.dataSchemaExpression.getValue(this.evaluationContext, message, URI.class))
+				.withSubject(this.subjectExpression.getValue(this.evaluationContext, message, String.class))
+				.withData(getPayload(message))
+				.withExtension(extensions)
+				.build();
+
+		return MessageBuilder.withPayload(eventFormat.serialize(cloudEvent))
+				.copyHeaders(message.getHeaders())
+				.build();
```

Comment:
We don't need to build a `Message` here. 
See `AbstractTransformer.transform()` code.